Update CONTRIBUTING.md

MuhammedgitAli opened the pull request:
**What this PR does / why we need it**:
Updates the Developer Guide to reflect the current state of Kubeflow Trainer V2. Provides more accurate instructions for setting up, building, and running the Trainer, including TODO notes for sections that need further refinement.

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/components/trainer/) included if any changes are user facing

tenzen-y reviewed (commented):
This actually seems to try focusing on removing old documentation.
I'm ok with that in this PR to avoid confusion.

@andreyvelich Any thoughts?

andreyvelich commented:
Let's not remove sections for the V1, but rather add the todo statements that they should be updated for V2.

andreyvelich commented on CONTRIBUTING.md:
```diff
@@ -1,8 +1,6 @@
-# TODO (andreyvelich): This doc needs to be updated for Kubeflow Trainer V2
-
-Kubeflow Training Operator is currently at v1.
+# TODO (Trainer V2 Update):  The testing and local Docker image building instructions need to be updated for Kubeflow Trainer V2. The current instructions in the repository are based on the Training Operator V1 and may not be directly applicable to the Trainer V2 project..
```
You can remove it since you can add TODO statement for each part which needs to be updated.

MuhammedgitAli replied:
Sounds good

andreyvelich commented on CONTRIBUTING.md:
```diff
@@ -1,8 +1,6 @@
-Kubeflow Training Operator is currently at v1.
+# TODO (Trainer V2 Update):  The testing and local Docker image building instructions need to be updated for Kubeflow Trainer V2. The current instructions in the repository are based on the Training Operator V1 and may not be directly applicable to the Trainer V2 project..
 
 ## Requirements
```
Let's update the Kind version from here: https://github.com/kubeflow/trainer/blob/master/go.mod

andreyvelich commented on CONTRIBUTING.md:
```diff
@@ -15,31 +13,6 @@ Kubeflow Training Operator is currently at v1.
 
-Note for Lima the link is to the Adopters, which supports several different container environments.
-
-## Building the operator
```
Actually, the following command should build images + deploy Kind cluster with your images.
Can you try to run it locally ?
```
make test-e2e-setup-cluster
```

MuhammedgitAli replied:
It worked , thank you !

andreyvelich commented on CONTRIBUTING.md:
```diff
@@ -15,31 +13,6 @@ Kubeflow Training Operator is currently at v1.
   - [Colima](https://github.com/abiosoft/colima) (Lima specifically for MacOS) (0.6.8 or later)
 - [pre-commit](https://pre-commit.com/)
 
-Note for Lima the link is to the Adopters, which supports several different container environments.
```
I think, it would be useful to explain in contributor guide on how to leverage our Makefile commands to develop Kubeflow Trainer project.
For example:
Run `make generate` to generate code assets.
Run `make test` to run Go unit tests
Run `make test-integration` to run Go integration tests.
etc..

andreyvelich replied:
Looking at other contributing guides, e.g:
- https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md

I like how Contributing guides are split between these sections:
### Development
Explain the development workflow and what tools are required to be installed to develop Kubeflow Trainer project.

### Testing
Explain the different testing we have in Kubeflow Trainer and how to run them:
- Unit test
- Integration test
- E2E test

### Best Practices

Coding Style, How we write code, etc.


I think, maybe we should structure our Contributing guide in a same way to make it clear for the new users how to contribute to Kubeflow Trainer.


@kubeflow/wg-training-leads @varodrig @franciscojavierarceo @Electronic-Waste @astefanutti @seanlaii @saileshd1402 Looking for your feedback and comments!

andreyvelich reviewed (commented):
Thank you for starting this @ryclondo!
I left a few comments, let's see what we can initially in this PR.

tenzen-y commented:
> Let's not remove sections for the V1, but rather add the todo statements that they should be updated for V2.

SGTM

MuhammedgitAli commented:
@andreyvelich @tenzen-y I restructured the CONTRIBUTING.md as suggested, organizing content into Development, Testing, and Best Practices sections. I also Updated Kind version to 0.27.0 and added instructions detailing the Makefile commands.

andreyvelich commented on CONTRIBUTING.md:
```diff
@@ -1,25 +1,67 @@
 
 - [Go](https://golang.org/) (1.23 or later)
-- [Docker](https://docs.docker.com/) (23 or later)
+- [Docker](https://www.docker.com/) (20.10.0 or later)
```
Why do you downgrade the Docker version ?
```suggestion
- [Docker](https://docs.docker.com/) (23 or later)
```

andreyvelich commented on CONTRIBUTING.md:
```diff
@@ -1,25 +1,67 @@
 - [kustomize](https://kustomize.io/) (4.0.5 or later)
-- [Kind](https://kind.sigs.k8s.io/) (0.22.0 or later)
+- [Kind](https://kind.sigs.k8s.io/) (0.27.0 or later)
 - [Lima](https://github.com/lima-vm/lima?tab=readme-ov-file#adopters) (an alternative to DockerDesktop) (0.21.0 or later)
   - [Colima](https://github.com/abiosoft/colima) (Lima specifically for MacOS) (0.6.8 or later)
```
Let's move Lima and Colima right after Docker since it is alternative to Docker Desktop
```suggestion
- [Docker]....
- [Lima](https://github.com/lima-vm/lima?tab=readme-ov-file#adopters) (an alternative to DockerDesktop) (0.21.0 or later)
  - [Colima](https://github.com/abiosoft/colima) (Lima specifically for MacOS) (0.6.8 or later)
```

andreyvelich commented on CONTRIBUTING.md:
```diff
@@ -1,25 +1,67 @@
-# TODO (andreyvelich): This doc needs to be updated for Kubeflow Trainer V2
-
-Kubeflow Training Operator is currently at v1.
+This guide explains how to contribute to the Kubeflow Trainer V2 project.
```
Please let's also add the link to the Kubeflow Trainer docs.
```suggestion
This guide explains how to contribute to the Kubeflow Trainer V2 project.
For the Kubeflow Trainer documentation, please check [the official Kubeflow documentation](https://www.kubeflow.org/docs/components/trainer/overview/).
```

andreyvelich commented on CONTRIBUTING.md:
```diff
@@ -1,25 +1,67 @@
+# Format and verify code
+make fmt vet golangci-lint
+
+# Run tests (see Testing section below)
+make test test-integration test-e2e
+
+# Setup a cluster for e2e testing
+make test-e2e-setup-cluster
```
Let's move these 2 command to Testing, since they are not required for Development.
```suggestion
```

andreyvelich commented on CONTRIBUTING.md:
```diff
@@ -1,25 +1,67 @@
+make generate
+
+# Download required tools for development
+make ginkgo envtest controller-gen kind
```
I think, this is not needed since, make generate contains instructions to install appropriate binaries.
 ```suggestion
```

tenzen-y replied:
This is not yet addressed.

andreyvelich replied:
@MuhammedgitAli Please can you address this comment also.

andreyvelich commented on CONTRIBUTING.md:
```diff
@@ -1,25 +1,67 @@
+3. Load your images into the cluster
+4. Deploy the operator
+
+#### Manual Setup
```
Please add the TODO statement that this part needs to be updated

andreyvelich commented on CONTRIBUTING.md:
````diff
@@ -1,25 +1,67 @@
+The simplest way to set up a local development environment is to use the provided Makefile command:
+
+```sh
+make test-e2e-setup-cluster
````
I think, this part can be moved under Testing section.

andreyvelich commented on CONTRIBUTING.md:
```diff
@@ -1,25 +1,50 @@
-## Building the operator
+## Development
+
+### Development Workflow
```
Can you make sections consistent and move all the legacy staff to the end of this doc ?
```
## Development
## Testing
## Best Practices
```

andreyvelich replied:
Hi @MuhammedgitAli, did you a chance to finish this PR so we can merge it, and start updating the other sections?

MuhammedgitAli replied:
Hi andrey, I just saw this pr. I fixed my prs reaching my email correctly as they were reaching my spam earlier.

MuhammedgitAli replied:
Hey andrey, I believe this pr is ready to be merged. I have the structure of the doc below if you would like a quick synopsis. 
# Developer Guide
## Requirements
## Development
    ### Development Workflow
## Testing
    ### Unit Tests
    ### Integration Tests
    ### E2E Tests
    ### Testing Local Changes
## Best Practices
    ### Go Version
    ### Code Style
## Legacy Setup Instructions
    #### Manual Setup
    ### Running the Operator Locally
    ### Running Sample Jobs
    ### Generate Python SDK Development

andreyvelich commented on CONTRIBUTING.md:
```diff
@@ -1,178 +1,269 @@
+The Kubeflow Trainer project includes several types of tests to ensure code quality and functionality.
 
-## Running the Operator Locally
+# Run tests 
+make test test-integration test-e2e
```
I think this section is duplicated with this: https://github.com/kubeflow/trainer/blob/6ce7b35fd43dbebb2724d7812a2036337429b28e/CONTRIBUTING.md#unit-tests

Can you move Unit test + Integration tests under `## testing`  section ?
```suggestion
```

MuhammedgitAli replied:
Hey andrey, I believe I already have Unit test + Integration tests under the Testing section in my most current pull request- are you perhaps looking at a previous pull request?

andreyvelich replied:
I think, I am looking at your latest commit, isn't?
https://github.com/kubeflow/trainer/pull/2512/commits/6818d6757c31fafbfc7a51fe4dc61b300bf5ce68
Right now, I can see that you updated it back to the previous stage.

Please can you update this doc according to the sections you mentioned in this comment: https://github.com/kubeflow/trainer/pull/2512#discussion_r1999835356 ?

andreyvelich replied:
Sorry my bad, I meant you just need to adjust the headers in this doc:
```
## Testing
### Unit Tests
....
### Integration Tests
...
### E2E Tests
...
```

Does it make sense ?

MuhammedgitAli replied:
Hey andrey, I think I have all of these in my commit number 6ce7b35

Here's the link: https://github.com/kubeflow/trainer/pull/2512/commits/6ce7b35fd43dbebb2724d7812a2036337429b28e
my bad for any misscommunication

andreyvelich replied:
That's right, but you need to remove the following text since it is duplicate in the E2E Tests section:
```

# Run tests 
make test test-integration test-e2e

# Setup a cluster for e2e testing
make test-e2e-setup-cluster


This command will:
1. Build the necessary Docker images
2. Create a Kind cluster if it doesn't exist
3. Load your images into the cluster
4. Deploy the operator

...

### Testing Local Changes

To test your local changes to the operator:

1. Build a custom operator image:
   ```sh
   make docker-build IMG=my-username/training-operator:my-pr-01
   ```

2. Load the image into your kind cluster:
   ```sh
   kind load docker-image my-username/training-operator:my-pr-01
   ```

3. Update the operator deployment to use your image:
   ```sh
   cd ./manifests/overlays/standalone
   kustomize edit set image my-username/training-operator=my-username/training-operator:my-pr-01
   ```

4. Deploy the updated operator:
   ```sh
   kubectl apply -k ./manifests/overlays/standalone
   ```

5. Submit a job to test:
   ```sh
   kubectl patch -n kubeflow deployments training-operator --type json -p '[{"op": "replace", "path": "/spec/template/spec/containers/0/image", "value": "my-username/training-operator:my-pr-01"}]'
   kubectl apply -f https://raw.githubusercontent.com/kubeflow/training-operator/master/examples/pytorch/simple.yaml
   ```

6. Check the job logs:
   ```sh
   kubectl logs -n kubeflow -l training.kubeflow.org/job-name=pytorch-simple
   ```

```

MuhammedgitAli replied:
I see what you meant now. Thanks for clarifying. I updated it here in 0c6469d

andreyvelich reviewed (commented):
Thank you for updating it @MuhammedgitAli!
I think, we should be good to merge this first change.
/lgtm
/assign @tenzen-y @Electronic-Waste @astefanutti

MuhammedgitAli commented:
Awesome! Thank you so much for the guidance during this pr.

MuhammedgitAli commented on CONTRIBUTING.md:
```diff
@@ -1,178 +1,219 @@
-# TODO (andreyvelich): This doc needs to be updated for Kubeflow Trainer V2
-
-Kubeflow Training Operator is currently at v1.
+This guide explains how to contribute to the Kubeflow Trainer V2 project.
```
I believe the legacy instructions are at the end, unless you were not referring to that?

astefanutti replied:
Oh I see. What about moving those in a separate file, or just point to the existing guide in the v1.x branch?

andreyvelich replied:
Actually, I can see that @MuhammedgitAli updated the majority of instructions for Kubeflow Trainer.
So we can just remote the V1 references from the file and create followup issues if we need to add more instructions to the contributing guide.
WDYT @MuhammedgitAli @astefanutti ?

andreyvelich replied:
Yeah, sorry, I meant remove the V1 references.

MuhammedgitAli replied:
Sounds great to me, I will get on it right away

Electronic-Waste commented on CONTRIBUTING.md:
```diff
@@ -1,178 +1,219 @@
+
+## Legacy Setup Instructions
+
+# TODO: The following setup instructions need to be updated for Kubeflow Trainer V2.
```
```suggestion
> TODO: The following setup instructions need to be updated for Kubeflow Trainer V2.
```

It looks weird:

<img width="1008" alt="截屏2025-03-20 00 19 28" src="https://github.com/user-attachments/assets/d005fcd8-f5b1-4989-8bf4-3c2837820978" />

MuhammedgitAli replied:
I will sign off on this if you are suggesting to use the > instead of the #.

andreyvelich replied:
Let's just use >, since we don't need to capitalize this TODO message.

Electronic-Waste commented on CONTRIBUTING.md:
Shall we add the running guide for v2 just like v1?

https://github.com/kubeflow/trainer/blob/5c89faa2c0f75fc23b0835286222f709cb9d30cf/CONTRIBUTING.md?plain=1#L43-L51

Electronic-Waste replied:
And also SDK Development:

https://github.com/kubeflow/trainer/blob/0c6469d851ead0cefd5bc01203258321d3d81b0c/CONTRIBUTING.md?plain=1#L212-L234

MuhammedgitAli replied:
@Electronic-Waste I am confused as to what you mean by running guide. Can you perhaps give an example of it like this in v1?

andreyvelich replied:
Let's update it as a followup PR.

Electronic-Waste reviewed (commented):
@MuhammedgitAli Thanks for this! Just a few comments:)

tenzen-y commented on CONTRIBUTING.md:
```diff
@@ -1,178 +1,219 @@
-# TODO (andreyvelich): This doc needs to be updated for Kubeflow Trainer V2
-
-Kubeflow Training Operator is currently at v1.
+This guide explains how to contribute to the Kubeflow Trainer V2 project.
```
SGTM

tenzen-y commented on CONTRIBUTING.md:
```diff
@@ -1,178 +1,219 @@
 - [Go](https://golang.org/) (1.23 or later)
 - [Docker](https://docs.docker.com/) (23 or later)
+- [Lima](https://github.com/lima-vm/lima?tab=readme-ov-file#adopters) (an alternative to DockerDesktop) (0.21.0 or later)
+- [Colima](https://github.com/abiosoft/colima) (Lima specifically for MacOS) (0.6.8 or later)
```
```suggestion
  - [Colima](https://github.com/abiosoft/colima) (Lima specifically for MacOS) (0.6.8 or later)
```

tenzen-y commented on CONTRIBUTING.md:
````diff
@@ -1,178 +1,219 @@
+To set up a Kind cluster for e2e testing:
 
+```sh
+make test-e2e-setup-cluster
````
This should be mentioned before `make test-e2e`

MuhammedgitAli replied:
@tenzen-y shall I remove the 

make generate

# Download required tools for development
make ginkgo envtest controller-gen kind

instructions then?

andreyvelich replied:
@MuhammedgitAli You just need make instructions for `E2E Tests` in this order:
```
To set up a Kind cluster for e2e testing:

make test-e2e-setup-cluster

Run the end-to-end tests with:

make test-e2e

```

MuhammedgitAli replied:
> This should be mentioned before `make test-e2e`

Sounds good. I moved it before `make test-e2e`

tenzen-y commented on CONTRIBUTING.md:
```diff
@@ -1,178 +1,219 @@
+On ubuntu the default go package appears to be gccgo-go which has problems. It's recommended to install Go from official tarballs.
 
-From here we can apply the manifests to the cluster.
+### Code Style
```
We could add similar things as https://github.com/kubeflow/katib/blob/master/CONTRIBUTING.md#go-development

MuhammedgitAli replied:
I think I have the same comments about pre commit and code style as in the katib contributing guide 
but just added an extra header for pre-commit as katib has that. Is there anything more you were referencing to add to add in Code style section that is similar to katib?

MuhammedgitAli replied:
@andreyvelich I can add this right away but to be sure, should I add a header below '##Code Style' that reads '###Best Practices" then have the below right under it?
When coding:

Follow [effective go](https://go.dev/doc/effective_go) guidelines.
Run locally [make check](https://github.com/kubeflow/katib/blob/46173463027e4fd2e604e25d7075b2b31a702049/Makefile#L31) to verify if changes follow best practices before submitting PRs.
Testing:

Use [cmp.Diff](https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff) instead of reflect.Equal, to provide useful comparisons.
Define test cases as maps instead of slices to avoid dependencies on the running order. Map key should be equal to the test case name.

andreyvelich replied:
@MuhammedgitAli Let's just rename the Go Version to Go Development section and add the following under it:
```
### Go Development

When coding:

Follow [effective go](https://go.dev/doc/effective_go) guidelines.
Run locally [make check](https://github.com/kubeflow/katib/blob/46173463027e4fd2e604e25d7075b2b31a702049/Makefile#L31) to verify if changes follow best practices before submitting PRs.
Testing:

Use [cmp.Diff](https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff) instead of reflect.Equal, to provide useful comparisons.
Define test cases as maps instead of slices to avoid dependencies on the running order. Map key should be equal to the test case name.
```

MuhammedgitAli replied:
Sounds good

andreyvelich commented:
@MuhammedgitAli @tenzen-y @Electronic-Waste Can we merge this PR, so @muzzlol] and @burhanuddin6 can work to address your changes in the followup PRs?
- https://github.com/kubeflow/trainer/pull/2561
- https://github.com/kubeflow/trainer/pull/2563

MuhammedgitAli commented:
@andreyvelich Is 04dbb64 ready to be merged as everything else as you mentioned will be handled in a follow-up pr ?

andreyvelich reviewed (commented):
Thank you for this great contribution @MuhammedgitAli!
/lgtm
/approve

tenzen-y commented on CONTRIBUTING.md:
```diff
@@ -1,178 +1,232 @@
+When coding:
+
+Follow [effective go](https://go.dev/doc/effective_go) guidelines.
+Run locally [make check](https://github.com/kubeflow/katib/blob/46173463027e4fd2e604e25d7075b2b31a702049/Makefile#L31) to verify if changes follow best practices before submitting PRs.
```
We want to fix this link.

muzzlol replied:
the link seem fine to me. although i might be missing what you're trying to point out
